### PR TITLE
Upstreaming v3 gh2

### PR DESCRIPTION
**Pull request by TheNetAdmin**
This is the second draft of our V3 patch set. I applied all code review comments from the first draft.

I'm running the `test-all-cases.sh` and will post the results later.

I'm also developing a new script to compare V2 and V3 outputs.

### CODE REVIEW

**Comment by marc-orr** [comment]
I've gone ahead and merged into a new branch, `upstreaming-v3-gh2`. I'll start taking a look at this revision. Thanks!

File: lib/efi.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#ifndef _EFI_H_
+#define _EFI_H_
+
+/*
+ * EFI-related functions in . This file's name "efi.h" is in
+ * conflict with GNU-EFI library's "efi.h", but  does not include
+ * GNU-EFI headers or links against GNU-EFI.
+ *
```
Comment (marc-orr) [inline lib/efi.h:8]: nit: Delete trailing whitespace: `* ` -> `*`.

File: lib/x86/setup.c [hunk 2]
```diff
@@ -254,6 +254,60 @@ efi_status_t setup_efi_pre_boot(unsigned long *mapkey, efi_bootinfo_t *efi_booti
...
+
+	/* Load 4-level page table */
+	write_cr3((ulong)&ptl4);
+}
+
+void setup_5level_page_table(void)
+{
+	/*
+	 * TODO: This function is a place holder for now. It is defined because
+	 * some test cases (e.g. x86/access.c) expects it to exist. If this
```
Comment (marc-orr) [inline lib/x86/setup.c:329]: nit: `expects` -> `expect`

File: lib/x86/setup.c [hunk 3]
```diff
@@ -254,6 +254,60 @@ efi_status_t setup_efi_pre_boot(unsigned long *mapkey, efi_bootinfo_t *efi_booti
...
+	 * function is not defined, gcc may generate wrong position-independent
+	 * code, which leads incorrect memory access: if compiling
```
Comment (marc-orr) [inline lib/x86/setup.c:331]: nit: `leads incorrect` -> `leads to incorrect`

File: lib/x86/setup.c [hunk 4]
```diff
@@ -254,6 +254,60 @@ efi_status_t setup_efi_pre_boot(unsigned long *mapkey, efi_bootinfo_t *efi_booti
...
+	 * x86/access.efi without this function defined, several data structures
+	 * (e.g. apic_ops) get compile time offset memory address, but they
```
Comment (marc-orr) [inline lib/x86/setup.c:333]: nit: `address` -> `addresses`

File: lib/x86/setup.c [hunk 5]
```diff
@@ -254,6 +254,60 @@ efi_status_t setup_efi_pre_boot(unsigned long *mapkey, efi_bootinfo_t *efi_booti
...
+	 * should get runtime %rip based address.
```
Comment (marc-orr) [inline lib/x86/setup.c:334]: nit: `address` -> `addresses`

**Comment by marc-orr** [comment]
Done with code review. Code looks good based on my review. I left some white space and grammar nits on two of the patches:
1. b4fcd39: x86 UEFI: Boot from UEFI
2. 586ba4f: x86 UEFI: Set up page tables

Also, I ran the amd_sev.img test case, and all test cases passed.

**Comment by TheNetAdmin** [comment]
The test and compare all cases results for this PR:

[test_logs_v3_gh2.zip](https://github.com/marc-orr/KVM-Unit-Tests-dev-fork/files/7274648/test_logs_v3_gh2.zip)

```
Comparing               access: diff result: 0
Comparing              asyncpf: diff result: 0
Comparing                  cet: diff result: 0
Comparing                debug: diff result: 0
Comparing             emulator: diff result: 0
Comparing            hypercall: diff result: 0
Comparing         hyperv_clock: diff result: 1
Comparing   hyperv_connections: diff result: 0
Comparing        hyperv_stimer: diff result: 0
Comparing         hyperv_synic: diff result: 0
Comparing             idt_test: diff result: 0
Comparing                 init: diff result: 0
Comparing          intel-iommu: diff result: 1
Comparing               ioapic: diff result: 0
Comparing        kvmclock_test: diff result: 1
Comparing               memory: diff result: 0
Comparing                  msr: diff result: 0
Comparing                 pcid: diff result: 0
Comparing                  pks: diff result: 0
Comparing                  pku: diff result: 0
Comparing                  pmu: diff result: 1
Comparing              pmu_lbr: diff result: 0
Comparing                rdpru: diff result: 0
Comparing           rmap_chain: diff result: 1
Comparing                   s3: diff result: 1
Comparing               setjmp: diff result: 0
Comparing                sieve: diff result: 0
Comparing                 smap: diff result: 1
Comparing              smptest: diff result: 0
Comparing              syscall: diff result: 0
Comparing                  tsc: diff result: 1
Comparing           tsc_adjust: diff result: 1
Comparing             tsx-ctrl: diff result: 0
Comparing                 umip: diff result: 0
Comparing               vmexit: diff result: 1
Comparing     vmware_backdoors: diff result: 0
Comparing                xsave: diff result: 0
```

I just developed a multi-threaded `test-all-cases.sh` to speedup the test runs. I'm working on the cross-version comparison script and will open another PR for these scripts.

Some test cases (e.g. `x86/pmu.c`) are sensitive to multi-threaded tests. Their results are not stable while multiple QEMUs are running at the same time:

```diff
# diff_log/pmu.diff
20c20
< PASS: ref cycles-2
---
> FAIL: ref cycles-2
26c26
< FAIL: llc misses-0
---
> PASS: llc misses-0
29c29
< FAIL: llc misses-3
---
> PASS: llc misses-3
55c55
< FAIL: all counters
---
> PASS: all counters
95c95
< PASS: full-width writes: llc misses-0
---
> FAIL: full-width writes: llc misses-0
124c124
< PASS: full-width writes: all counters
---
> FAIL: full-width writes: all counters
```

**Comment by TheNetAdmin** [comment]
V2 and V3-gh2 comparison:

[logs-v2-v3.zip](https://github.com/marc-orr/KVM-Unit-Tests-dev-fork/files/7274716/logs-v2-v3.zip)

```
Comparing:             access.efi.uefi.filtered.log: diff result 0
Comparing:         access.flat.seabios.filtered.log: diff result 0
Comparing:            asyncpf.efi.uefi.filtered.log: diff result 0
Comparing:        asyncpf.flat.seabios.filtered.log: diff result 0
Comparing:                cet.efi.uefi.filtered.log: diff result 0
Comparing:            cet.flat.seabios.filtered.log: diff result 0
Comparing:              debug.efi.uefi.filtered.log: diff result 0
Comparing:          debug.flat.seabios.filtered.log: diff result 0
Comparing:           emulator.efi.uefi.filtered.log: diff result 0
Comparing:       emulator.flat.seabios.filtered.log: diff result 0
Comparing:          hypercall.efi.uefi.filtered.log: diff result 0
Comparing:      hypercall.flat.seabios.filtered.log: diff result 0
Comparing:       hyperv_clock.efi.uefi.filtered.log: diff result 1
Comparing:   hyperv_clock.flat.seabios.filtered.log: diff result 1
Comparing: hyperv_connections.efi.uefi.filtered.log: diff result 0
Comparing: hyperv_connections.flat.seabios.filtered.log: diff result 0
Comparing:      hyperv_stimer.efi.uefi.filtered.log: diff result 0
Comparing:  hyperv_stimer.flat.seabios.filtered.log: diff result 0
Comparing:       hyperv_synic.efi.uefi.filtered.log: diff result 0
Comparing:   hyperv_synic.flat.seabios.filtered.log: diff result 0
Comparing:           idt_test.efi.uefi.filtered.log: diff result 0
Comparing:       idt_test.flat.seabios.filtered.log: diff result 0
Comparing:               init.efi.uefi.filtered.log: diff result 0
Comparing:           init.flat.seabios.filtered.log: diff result 0
Comparing:        intel-iommu.efi.uefi.filtered.log: diff result 0
Comparing:    intel-iommu.flat.seabios.filtered.log: diff result 0
Comparing:             ioapic.efi.uefi.filtered.log: diff result 0
Comparing:         ioapic.flat.seabios.filtered.log: diff result 0
Comparing:      kvmclock_test.efi.uefi.filtered.log: diff result 1
Comparing:  kvmclock_test.flat.seabios.filtered.log: diff result 1
Comparing:             memory.efi.uefi.filtered.log: diff result 0
Comparing:         memory.flat.seabios.filtered.log: diff result 0
Comparing:                msr.efi.uefi.filtered.log: diff result 0
Comparing:            msr.flat.seabios.filtered.log: diff result 0
Comparing:               pcid.efi.uefi.filtered.log: diff result 0
Comparing:           pcid.flat.seabios.filtered.log: diff result 0
Comparing:                pks.efi.uefi.filtered.log: diff result 0
Comparing:            pks.flat.seabios.filtered.log: diff result 0
Comparing:                pku.efi.uefi.filtered.log: diff result 0
Comparing:            pku.flat.seabios.filtered.log: diff result 0
Comparing:                pmu.efi.uefi.filtered.log: diff result 1
Comparing:            pmu.flat.seabios.filtered.log: diff result 1
Comparing:            pmu_lbr.efi.uefi.filtered.log: diff result 0
Comparing:        pmu_lbr.flat.seabios.filtered.log: diff result 0
Comparing:              rdpru.efi.uefi.filtered.log: diff result 0
Comparing:          rdpru.flat.seabios.filtered.log: diff result 0
Comparing:         rmap_chain.efi.uefi.filtered.log: diff result 0
Comparing:     rmap_chain.flat.seabios.filtered.log: diff result 0
Comparing:                 s3.efi.uefi.filtered.log: diff result 0
Comparing:             s3.flat.seabios.filtered.log: diff result 1
Comparing:             setjmp.efi.uefi.filtered.log: diff result 0
Comparing:         setjmp.flat.seabios.filtered.log: diff result 0
Comparing:              sieve.efi.uefi.filtered.log: diff result 0
Comparing:          sieve.flat.seabios.filtered.log: diff result 0
Comparing:               smap.efi.uefi.filtered.log: diff result 0
Comparing:           smap.flat.seabios.filtered.log: diff result 0
Comparing:            smptest.efi.uefi.filtered.log: diff result 0
Comparing:        smptest.flat.seabios.filtered.log: diff result 0
Comparing:            syscall.efi.uefi.filtered.log: diff result 0
Comparing:        syscall.flat.seabios.filtered.log: diff result 0
Comparing:         tsc_adjust.efi.uefi.filtered.log: diff result 1
Comparing:     tsc_adjust.flat.seabios.filtered.log: diff result 0
Comparing:                tsc.efi.uefi.filtered.log: diff result 1
Comparing:            tsc.flat.seabios.filtered.log: diff result 1
Comparing:           tsx-ctrl.efi.uefi.filtered.log: diff result 0
Comparing:       tsx-ctrl.flat.seabios.filtered.log: diff result 0
Comparing:               umip.efi.uefi.filtered.log: diff result 0
Comparing:           umip.flat.seabios.filtered.log: diff result 0
Comparing:             vmexit.efi.uefi.filtered.log: diff result 0
Comparing:         vmexit.flat.seabios.filtered.log: diff result 1
Comparing:   vmware_backdoors.efi.uefi.filtered.log: diff result 0
Comparing: vmware_backdoors.flat.seabios.filtered.log: diff result 0
Comparing:              xsave.efi.uefi.filtered.log: diff result 0
Comparing:          xsave.flat.seabios.filtered.log: diff result 0
```

There are some discrepancies. I think they are all performance-related fluctuations, not code logic bugs. So I think we are good to go.

I will open a new PR to fix the grammar nits.

**Comment by marc-orr** [comment]
> V2 and V3-gh2 comparison:
> 
> [logs-v2-v3.zip](https://github.com/marc-orr/KVM-Unit-Tests-dev-fork/files/7274716/logs-v2-v3.zip)
> 
> ```
> Comparing:             access.efi.uefi.filtered.log: diff result 0
> Comparing:         access.flat.seabios.filtered.log: diff result 0
> Comparing:            asyncpf.efi.uefi.filtered.log: diff result 0
> Comparing:        asyncpf.flat.seabios.filtered.log: diff result 0
> Comparing:                cet.efi.uefi.filtered.log: diff result 0
> Comparing:            cet.flat.seabios.filtered.log: diff result 0
> Comparing:              debug.efi.uefi.filtered.log: diff result 0
> Comparing:          debug.flat.seabios.filtered.log: diff result 0
> Comparing:           emulator.efi.uefi.filtered.log: diff result 0
> Comparing:       emulator.flat.seabios.filtered.log: diff result 0
> Comparing:          hypercall.efi.uefi.filtered.log: diff result 0
> Comparing:      hypercall.flat.seabios.filtered.log: diff result 0
> Comparing:       hyperv_clock.efi.uefi.filtered.log: diff result 1
> Comparing:   hyperv_clock.flat.seabios.filtered.log: diff result 1
> Comparing: hyperv_connections.efi.uefi.filtered.log: diff result 0
> Comparing: hyperv_connections.flat.seabios.filtered.log: diff result 0
> Comparing:      hyperv_stimer.efi.uefi.filtered.log: diff result 0
> Comparing:  hyperv_stimer.flat.seabios.filtered.log: diff result 0
> Comparing:       hyperv_synic.efi.uefi.filtered.log: diff result 0
> Comparing:   hyperv_synic.flat.seabios.filtered.log: diff result 0
> Comparing:           idt_test.efi.uefi.filtered.log: diff result 0
> Comparing:       idt_test.flat.seabios.filtered.log: diff result 0
> Comparing:               init.efi.uefi.filtered.log: diff result 0
> Comparing:           init.flat.seabios.filtered.log: diff result 0
> Comparing:        intel-iommu.efi.uefi.filtered.log: diff result 0
> Comparing:    intel-iommu.flat.seabios.filtered.log: diff result 0
> Comparing:             ioapic.efi.uefi.filtered.log: diff result 0
> Comparing:         ioapic.flat.seabios.filtered.log: diff result 0
> Comparing:      kvmclock_test.efi.uefi.filtered.log: diff result 1
> Comparing:  kvmclock_test.flat.seabios.filtered.log: diff result 1
> Comparing:             memory.efi.uefi.filtered.log: diff result 0
> Comparing:         memory.flat.seabios.filtered.log: diff result 0
> Comparing:                msr.efi.uefi.filtered.log: diff result 0
> Comparing:            msr.flat.seabios.filtered.log: diff result 0
> Comparing:               pcid.efi.uefi.filtered.log: diff result 0
> Comparing:           pcid.flat.seabios.filtered.log: diff result 0
> Comparing:                pks.efi.uefi.filtered.log: diff result 0
> Comparing:            pks.flat.seabios.filtered.log: diff result 0
> Comparing:                pku.efi.uefi.filtered.log: diff result 0
> Comparing:            pku.flat.seabios.filtered.log: diff result 0
> Comparing:                pmu.efi.uefi.filtered.log: diff result 1
> Comparing:            pmu.flat.seabios.filtered.log: diff result 1
> Comparing:            pmu_lbr.efi.uefi.filtered.log: diff result 0
> Comparing:        pmu_lbr.flat.seabios.filtered.log: diff result 0
> Comparing:              rdpru.efi.uefi.filtered.log: diff result 0
> Comparing:          rdpru.flat.seabios.filtered.log: diff result 0
> Comparing:         rmap_chain.efi.uefi.filtered.log: diff result 0
> Comparing:     rmap_chain.flat.seabios.filtered.log: diff result 0
> Comparing:                 s3.efi.uefi.filtered.log: diff result 0
> Comparing:             s3.flat.seabios.filtered.log: diff result 1
> Comparing:             setjmp.efi.uefi.filtered.log: diff result 0
> Comparing:         setjmp.flat.seabios.filtered.log: diff result 0
> Comparing:              sieve.efi.uefi.filtered.log: diff result 0
> Comparing:          sieve.flat.seabios.filtered.log: diff result 0
> Comparing:               smap.efi.uefi.filtered.log: diff result 0
> Comparing:           smap.flat.seabios.filtered.log: diff result 0
> Comparing:            smptest.efi.uefi.filtered.log: diff result 0
> Comparing:        smptest.flat.seabios.filtered.log: diff result 0
> Comparing:            syscall.efi.uefi.filtered.log: diff result 0
> Comparing:        syscall.flat.seabios.filtered.log: diff result 0
> Comparing:         tsc_adjust.efi.uefi.filtered.log: diff result 1
> Comparing:     tsc_adjust.flat.seabios.filtered.log: diff result 0
> Comparing:                tsc.efi.uefi.filtered.log: diff result 1
> Comparing:            tsc.flat.seabios.filtered.log: diff result 1
> Comparing:           tsx-ctrl.efi.uefi.filtered.log: diff result 0
> Comparing:       tsx-ctrl.flat.seabios.filtered.log: diff result 0
> Comparing:               umip.efi.uefi.filtered.log: diff result 0
> Comparing:           umip.flat.seabios.filtered.log: diff result 0
> Comparing:             vmexit.efi.uefi.filtered.log: diff result 0
> Comparing:         vmexit.flat.seabios.filtered.log: diff result 1
> Comparing:   vmware_backdoors.efi.uefi.filtered.log: diff result 0
> Comparing: vmware_backdoors.flat.seabios.filtered.log: diff result 0
> Comparing:              xsave.efi.uefi.filtered.log: diff result 0
> Comparing:          xsave.flat.seabios.filtered.log: diff result 0
> ```
> 
> There are some discrepancies. I think they are all performance-related fluctuations, not code logic bugs. So I think we are good to go.
> 
> I will open a new PR to fix the grammar nits.

Sounds perfect. I think we can post `upstreaming-v3-gh3` to the mailing list. We will need to annotate the cover letter with a summary of the v2 -> v3 changes before posting v3 to mailing list. Here's a good example on how to do that. See the `v1 -> v2` section in the cover letter. And under that, for each patch that changed, he's got the patch number (maybe patch title is better?) and a summary of what changed since the last revision.

https://lore.kernel.org/kvm/20210604172611.281819-1-jmattson@google.com/

**Comment by TheNetAdmin** [comment]
> Sounds perfect. I think we can post `upstreaming-v3-gh3` to the mailing list. We will need to annotate the cover letter with a summary of the v2 -> v3 changes before posting v3 to mailing list. Here's a good example on how to do that. See the `v1 -> v2` section in the cover letter. And under that, for each patch that changed, he's got the patch number (maybe patch title is better?) and a summary of what changed since the last revision.
> 
> https://lore.kernel.org/kvm/20210604172611.281819-1-jmattson@google.com/

This is a great example! Let me first figure out how to set up git email client on my workstation, and then I will draft a V3 cover letter.

**Comment by marc-orr** [comment]
Sounds good. FYI, all the `git send-email` scripts you were using should be
in the `dev` branch.

**Pull request merged**